File: prakash_steel/prakash_steel/report/tog_calculation_report/tog_calculation_report.py
import math
from collections import defaultdict

import frappe
from frappe import _
from frappe.utils import cint, flt, today, add_days

from prakash_steel.utils.lead_time import get_default_bom


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data


@frappe.whitelist()
def get_tog_parent_sell_console_payload():
	"""
	Return parent-sell debug JSON for the browser console.

	Loaded via a separate request so it is not lost when the report pipeline
	(e.g. user-permission filtering on Item rows) strips or reorders rows.
	"""
	if not frappe.has_permission("Item", "report"):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	rows = get_data(None)
	if not rows:
		return {}
	return rows[0].get("tog_parent_sell_debug") or {}


@frappe.whitelist()
def get_item_tog_debug(item_code):
	"""
	Full per-item debug for browser console.

	Returns:
	  - direct_sells     : every Sales Invoice line that sold this item (→ Sell column)
	  - parent_sell_contributions : every BOM hop that pushed demand onto this item (→ Parent Sell column)
	  - total_sell / total_parent_sell_raw / total_parent_sell_ceiled : summary numbers
	"""
	if not frappe.has_permission("Item", "report"):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	days = _get_horizon_days()
	sales_lines = _get_sales_invoice_lines()

	# ── Direct sells (Sell column) ─────────────────────────────────────────
	direct_sells = [l for l in sales_lines if l.get("item_code") == item_code]
	total_sell = sum(flt(l.get("qty", 0)) for l in direct_sells)

	# ── Propagation trace ──────────────────────────────────────────────────
	trace = _compute_parent_sell_from_lines(sales_lines, collect_trace=True)[1]

	# Only hops where demand lands ON this item
	contributions = [t for t in (trace or []) if t.get("to_item") == item_code]

	# parent_sell accumulator = sum of path_total_accrued_to_to_item (same as acc[item_code])
	total_parent_sell_raw = sum(flt(t.get("path_total_accrued_to_to_item", 0)) for t in contributions)

	return {
		"item_code": item_code,
		"horizon_days": days,
		"total_sell": int(total_sell),
		"total_parent_sell_raw": total_parent_sell_raw,
		"total_parent_sell_ceiled": int(math.ceil(total_parent_sell_raw)) if total_parent_sell_raw > 0 else 0,
		"direct_sells": direct_sells,
		"parent_sell_contributions": contributions,
	}


def get_columns():
	"""Report column definitions."""
	return [
		{
			"label": "Item Code",
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 140,
		},
		{
			"label": "Category Name",
			"fieldname": "custom_category_name",
			"fieldtype": "Data",
			"width": 150,
		},
		{
			"label": "Item Type",
			"fieldname": "custom_item_type",
			"fieldtype": "Data",
			"width": 120,
		},
		{
			"label": "Sell",
			"fieldname": "sell",
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"label": "Parent Sell",
			"fieldname": "parent_sell",
			"fieldtype": "Int",
			"width": 110,
		},
		{
			"label": "Consumption",
			"fieldname": "consumption",
			"fieldtype": "Int",
			"width": 110,
			"hidden": 1,  # Not shown by default since it's not used in ADU calculation for all items; can be toggled on if needed
		},
		{
			"label": "ADU",
			"fieldname": "custom_adu",
			"fieldtype": "Int",
			"width": 90,
		},
		{
			"label": "SD",
			"fieldname": "sd",
			"fieldtype": "Float",
			"width": 90,
		},
		{
			"label": "COV",
			"fieldname": "cov",
			"fieldtype": "Float",
			"width": 90,
		},
	]


def get_data(filters=None):
	"""
	Get data from Item doctype.

	Populates:
	- item_code (from Item.name)
	- custom_category_name
	- custom_item_type
	- custom_adu

	Leaves other fields (consumption, sd, cov) empty.

	Debug payload is stored on the first row as `tog_parent_sell_debug` and also
	exposed via `get_tog_parent_sell_console_payload` (used by report JS) so it
	survives row filtering in `generate_report_result`.
	"""
	# Number of days in horizon (based on ADU Horizon)
	days = _get_horizon_days()
	# Map of item_code -> total sales qty in horizon
	sales_qty_map = _get_sales_qty_by_item()
	# Map of item_code -> total consumption (filtered by same date range as sales)
	consumption_map = _get_consumption_by_item(days)

	sales_lines = _get_sales_invoice_lines()
	parent_sell_map, propagation_trace = _compute_parent_sell_from_lines(
		sales_lines, collect_trace=True
	)

	items = frappe.db.get_all(
		"Item",
		fields=[
			"name as item_code",
			"custom_category_name",
			"custom_item_type",
			"custom_adu",
		],
		# Adjust filters if you want to limit which items appear
		# Example: only stock items, not disabled
		# filters={"is_stock_item": 1, "disabled": 0},
	)

	# Populate measure columns
	for item in items:
		item_code = item.get("item_code")

		# 1) Sell: live total sales qty in current horizon (integer)
		sell_qty = flt(sales_qty_map.get(item_code, 0.0))
		item["sell"] = int(sell_qty) if sell_qty else 0

		# 1b) Parent sell: demand on this item from exploding other items' sales through BOMs (ceiled int)
		ps = flt(parent_sell_map.get(item_code, 0.0))
		item["parent_sell"] = int(math.ceil(ps)) if ps > 0 else 0

		# 2) Consumption: live total consumption (integer)
		consumption_qty = flt(consumption_map.get(item_code, 0.0))
		item["consumption"] = int(consumption_qty) if consumption_qty else 0

		# 3) ADU: always recalculate from (Sell + Consumption) / days (ceiled to whole number)
		total_usage = sell_qty + consumption_qty
		if days > 0 and total_usage > 0:
			adu_raw = total_usage / days
			item["custom_adu"] = math.ceil(adu_raw)
		else:
			# If horizon not configured or no sales/consumption, set ADU to 0
			item["custom_adu"] = 0

		# 4) SD / COV placeholders (not yet implemented)
		item.setdefault("sd", None)
		item.setdefault("cov", None)

	# Browser console: one payload on first row (always present in API result; not a column).
	if items:
		per_item = []
		for row in items:
			ic = row.get("item_code")
			ps = flt(parent_sell_map.get(ic, 0.0))
			per_item.append(
				{
					"item_code": ic,
					"parent_sell_float": ps,
					"parent_sell_in_grid": int(math.ceil(ps)) if ps > 0 else 0,
					"sell_in_grid": row.get("sell"),
				}
			)
		items[0]["tog_parent_sell_debug"] = {
			"horizon_days": days,
			"sales_invoice_lines": sales_lines,
			"propagation_trace": propagation_trace or [],
			"per_item_parent_sell": per_item,
		}

	return items


def _get_horizon_days() -> int:
	"""Return number of days to look back based on ADU Horizon single DocType."""
	horizon = frappe.get_single("ADU Horizon")
	weeks_raw = getattr(horizon, "week", None) or 0
	weeks = cint(weeks_raw) or 0
	if weeks <= 0:
		return 0
	return weeks * 7


def _get_sales_qty_by_item() -> dict[str, float]:
	"""
	Return a mapping of item_code -> total sales quantity
	within the ADU horizon (same range used in ADU calculation),
	but WITHOUT dividing by days/weeks.
	"""
	days = _get_horizon_days()
	if days <= 0:
		return {}

	end_date = today()
	# Include "days" number of days including today
	start_date = add_days(end_date, -(days - 1))

	rows = frappe.db.sql(
		"""
		SELECT
			sii.item_code,
			COALESCE(SUM(sii.qty), 0) AS total_qty
		FROM `tabSales Invoice Item` sii
		INNER JOIN `tabSales Invoice` si ON si.name = sii.parent
		WHERE
			si.docstatus = 1
			AND si.posting_date BETWEEN %s AND %s
		GROUP BY sii.item_code
		""",
		(start_date, end_date),
		as_dict=True,
	)

	return {row["item_code"]: flt(row.get("total_qty", 0.0)) for row in rows}


def _get_sales_invoice_lines() -> list[dict]:
	"""Each submitted Sales Invoice line in the ADU horizon (one row per line item)."""
	days = _get_horizon_days()
	if days <= 0:
		return []

	end_date = today()
	start_date = add_days(end_date, -(days - 1))

	rows = frappe.db.sql(
		"""
		SELECT
			sii.name AS invoice_item_row,
			sii.parent AS sales_invoice,
			si.posting_date AS posting_date,
			sii.item_code AS item_code,
			sii.qty AS qty,
			sii.sales_order AS sales_order
		FROM `tabSales Invoice Item` sii
		INNER JOIN `tabSales Invoice` si ON si.name = sii.parent
		WHERE
			si.docstatus = 1
			AND si.posting_date BETWEEN %s AND %s
		ORDER BY si.posting_date, sii.parent, sii.idx
		""",
		(start_date, end_date),
		as_dict=True,
	)

	out = []
	for row in rows:
		out.append(
			{
				"invoice_item_row": row.get("invoice_item_row"),
				"sales_invoice": row.get("sales_invoice"),
				"posting_date": str(row.get("posting_date") or ""),
				"item_code": row.get("item_code"),
				"qty": flt(row.get("qty")),
				"sales_order": row.get("sales_order") or None,
			}
		)
	return out


def _compute_parent_sell_map(sales_qty_map: dict[str, float]) -> dict[str, float]:
	"""Aggregated by item (no per–sales-invoice trace)."""
	lines = [
		{
			"invoice_item_row": None,
			"sales_invoice": None,
			"posting_date": None,
			"item_code": k,
			"qty": flt(v),
			"sales_order": None,
		}
		for k, v in sales_qty_map.items()
		if flt(v) > 0
	]
	acc, _ = _compute_parent_sell_from_lines(lines, collect_trace=False)
	return acc


def _compute_parent_sell_from_lines(
	lines: list[dict],
	collect_trace: bool,
) -> tuple[dict[str, float], list | None]:
	"""
	For each sales line, explode through default BOMs.

	Per BOM row: child_need = incoming_qty * (bom.quantity / bom_item.qty).
	Parent sell for a component is the cumulative demand along the path.

	If collect_trace, each hop is appended with sales invoice / SO context.
	"""
	acc = defaultdict(float)
	trace: list[dict] | None = [] if collect_trace else None
	bom_doc_cache: dict[str, object] = {}
	default_bom_cache: dict[str, str | None] = {}
	rm_cache: dict[str, bool] = {}

	def _is_rm_stop(item_code: str) -> bool:
		if item_code not in rm_cache:
			raw = frappe.db.get_value("Item", item_code, "custom_item_type")
			rm_cache[item_code] = bool(raw and str(raw).strip().lower() == "rm")
		return rm_cache[item_code]

	def _get_bom_doc(bom_name: str):
		if bom_name not in bom_doc_cache:
			bom_doc_cache[bom_name] = frappe.get_doc("BOM", bom_name)
		return bom_doc_cache[bom_name]

	def _propagate(
		item_code: str,
		incoming_qty: float,
		visited: frozenset[str],
		path_cumulative: float,
		line_ctx: dict,
	) -> None:
		if not incoming_qty or item_code in visited:
			return
		if _is_rm_stop(item_code):
			return

		if item_code not in default_bom_cache:
			default_bom_cache[item_code] = get_default_bom(item_code)
		bom_name = default_bom_cache[item_code]
		if not bom_name:
			return

		bom_doc = _get_bom_doc(bom_name)
		fg_qty = flt(bom_doc.quantity)
		if fg_qty <= 0:
			return

		next_visited = visited | {item_code}

		for bom_item in bom_doc.items:
			child_code = bom_item.item_code
			child_line_qty = flt(bom_item.qty)
			if not child_code or child_line_qty <= 0:
				continue
			child_need = incoming_qty * (fg_qty / child_line_qty)
			path_total = path_cumulative + child_need
			acc[child_code] += path_total

			if trace is not None:
				trace.append(
					{
						"sales_invoice": line_ctx.get("sales_invoice"),
						"sales_order": line_ctx.get("sales_order"),
						"invoice_item_row": line_ctx.get("invoice_item_row"),
						"posting_date": line_ctx.get("posting_date"),
						"sold_item": line_ctx.get("item_code"),
						"sold_qty_this_line": flt(line_ctx.get("qty")),
						"from_item": item_code,
						"to_item": child_code,
						"bom": bom_name,
						"bom_fg_qty": fg_qty,
						"bom_line_qty": child_line_qty,
						"child_need": child_need,
						"path_cumulative_before_hop": path_cumulative,
						"path_total_accrued_to_to_item": path_total,
					}
				)

			if _is_rm_stop(child_code):
				continue
			_propagate(child_code, child_need, next_visited, path_total, line_ctx)

	for line in lines:
		sq = flt(line.get("qty"))
		if sq <= 0 or not line.get("item_code"):
			continue
		line_ctx = {
			"invoice_item_row": line.get("invoice_item_row"),
			"sales_invoice": line.get("sales_invoice"),
			"posting_date": line.get("posting_date"),
			"item_code": line.get("item_code"),
			"qty": sq,
			"sales_order": line.get("sales_order"),
		}
		_propagate(line["item_code"], sq, frozenset(), 0.0, line_ctx)

	return dict(acc), trace


def _get_consumption_by_item(days: int) -> dict[str, float]:
	"""
	Return a mapping of item_code -> total consumption within the ADU horizon.

	Logic:
	- For items with item_type = "rb" or "bo": sum actual_rm_consumption from Bright Bar Production doctype
	- For items with item_type = "rm" or "traded": sum billet_weight from Billet Cutting doctype
	- Filtered by the same date range as sales (ADU horizon)
	"""
	consumption_map = {}

	if days <= 0:
		return consumption_map

	end_date = today()
	# Include "days" number of days including today
	start_date = add_days(end_date, -(days - 1))

	# Get consumption from Bright Bar Production for items with item_type = "rb" or "bo"

	bright_bar_rows = frappe.db.sql(
		"""
		SELECT
			bbp.raw_material AS item_code,
			COALESCE(SUM(bbp.actual_rm_consumption), 0) AS total_rm_consumption
		FROM `tabBright Bar Production` bbp
		INNER JOIN `tabItem` i ON i.name = bbp.raw_material
		WHERE
			bbp.docstatus = 1
			AND LOWER(i.custom_item_type) IN ('rb', 'bo')
			AND bbp.production_date BETWEEN %s AND %s
		GROUP BY bbp.raw_material
		""",
		(start_date, end_date),
		as_dict=True,
	)

	for row in bright_bar_rows:
		item_code = row.get("item_code")
		if item_code:
			consumption_map[item_code] = flt(row.get("total_rm_consumption", 0.0))

	# Get consumption from Billet Cutting for items with item_type = "rm" or "traded"
	billet_cutting_rows = frappe.db.sql(
		"""
		SELECT
			bc.billet_size AS item_code,
			COALESCE(SUM(bc.billet_weight), 0) AS total_billet_weight
		FROM `tabBillet Cutting` bc
		INNER JOIN `tabItem` i ON i.name = bc.billet_size
		WHERE
			bc.docstatus = 1
			AND LOWER(i.custom_item_type) IN ('rm', 'traded')
			AND bc.posting_date BETWEEN %s AND %s
		GROUP BY bc.billet_size
		""",
		(start_date, end_date),
		as_dict=True,
	)

	for row in billet_cutting_rows:
		item_code = row.get("item_code")
		if item_code:
			# If item already has consumption from Bright Bar Production, add to it
			# Otherwise, set it
			current_consumption = consumption_map.get(item_code, 0.0)
			consumption_map[item_code] = current_consumption + flt(row.get("total_billet_weight", 0.0))

	return consumption_map

File: prakash_steel/prakash_steel/report/tog_calculation_report/tog_calculation_report.js
// Copyright (c) 2026, beetashok chakraborty and contributors
// For license information, please see license.txt

const TOG_ITEM_DEBUG_METHOD =
	"prakash_steel.prakash_steel.report.tog_calculation_report.tog_calculation_report.get_item_tog_debug";

function tog_debug_item(item_code) {
	if (!item_code) return;
	console.info(`[Tog] Fetching debug for: ${item_code} …`);

	frappe.call({
		method: TOG_ITEM_DEBUG_METHOD,
		args: { item_code },
		callback: function (r) {
			const d = r && r.message;
			if (!d) {
				console.warn("[Tog] No data returned", r);
				return;
			}

			// ── Header ──────────────────────────────────────────────────────
			console.group(
				`%c[Tog] ═══ Debug: ${d.item_code} ═══`,
				"font-size:14px; font-weight:bold; color:#2196F3"
			);
			console.log(`Horizon: ${d.horizon_days} days`);
			console.log(
				`%cSell (direct)  = ${d.total_sell}`,
				"font-weight:bold; color:#4CAF50"
			);
			console.log(
				`%cParent Sell    = ${d.total_parent_sell_raw.toFixed(4)}  →  ceiled in grid = ${d.total_parent_sell_ceiled}`,
				"font-weight:bold; color:#FF9800"
			);

			// ── Section 1: Direct Sells ──────────────────────────────────────
			console.group(
				`%c[SELL] "${d.item_code}" was directly sold on ${d.direct_sells.length} invoice line(s)`,
				"color:#4CAF50; font-weight:bold"
			);
			if (d.direct_sells.length) {
				console.table(
					d.direct_sells.map((l) => ({
						Date: l.posting_date,
						"Sales Invoice": l.sales_invoice,
						"Sales Order": l.sales_order || "-",
						Qty: l.qty,
					}))
				);
			} else {
				console.log("No direct sales in this horizon.");
			}
			console.groupEnd();

			// ── Section 2: Parent Sell contributions ─────────────────────────
			console.group(
				`%c[PARENT SELL] ${d.parent_sell_contributions.length} BOM hop(s) pushed demand onto "${d.item_code}"`,
				"color:#FF9800; font-weight:bold"
			);
			console.log(
				"Formula per hop:  child_need = sold_qty × (bom_fg_qty ÷ bom_line_qty)\n" +
				"  path_total_added_to_parent_sell = path_cumul_before + child_need  (accumulates along BOM depth)"
			);

			if (d.parent_sell_contributions.length) {
				// Group by invoice line so each sale event is one collapsible block
				const groups = {};
				for (const hop of d.parent_sell_contributions) {
					const key = `${hop.posting_date}  |  ${hop.sales_invoice}  |  Sold: ${hop.sold_item}  qty=${hop.sold_qty_this_line}`;
					if (!groups[key]) groups[key] = [];
					groups[key].push(hop);
				}

				for (const [label, hops] of Object.entries(groups)) {
					// Show BOM ratio in the label so user sees it without expanding
					const h0 = hops[0];
					const ratio = h0.bom_fg_qty / h0.bom_line_qty;
					const labelWithRatio = `${label}  →  BOM ratio ${h0.bom_fg_qty}÷${h0.bom_line_qty}=${ratio.toFixed(4)}  →  child_need = sold_qty × ${ratio.toFixed(4)}`;
					console.groupCollapsed(labelWithRatio);
					console.table(
						hops.map((h) => ({
							"BOM parent (from)": h.from_item,
							"This item (to)": h.to_item,
							BOM: h.bom,
							"bom_fg_qty (A)": h.bom_fg_qty,
							"bom_line_qty (B)": h.bom_line_qty,
							"ratio A/B": +(h.bom_fg_qty / h.bom_line_qty).toFixed(4),
							"sold_qty (Q)": h.sold_qty_this_line,
							"child_need = Q×(A/B)": +h.child_need.toFixed(4),
							"path_cumul_before": +h.path_cumulative_before_hop.toFixed(4),
							"path_total → added to Parent Sell": +h.path_total_accrued_to_to_item.toFixed(4),
						}))
					);
					console.groupEnd();
				}

				// Summary: which sold item contributed how much
				const by_sold = {};
				for (const hop of d.parent_sell_contributions) {
					by_sold[hop.sold_item] = (by_sold[hop.sold_item] || 0) + hop.path_total_accrued_to_to_item;
				}
				console.group("Contribution summary by sold item");
				console.table(
					Object.entries(by_sold).map(([item, total]) => ({
						"Sold Item": item,
						"Total added to Parent Sell": +total.toFixed(4),
					}))
				);
				console.groupEnd();
			} else {
				console.log(
					"No BOM hops found. Possible reasons:\n" +
					"  • No parent item sold in this horizon has a default BOM that includes this item\n" +
					"  • This item has item_type = RM (stop condition — BOM traversal stops here)"
				);
			}

			console.groupEnd(); // PARENT SELL
			console.groupEnd(); // main group
		},
		error: function (err) {
			console.error(`[Tog] Request failed for "${item_code}"`, err);
		},
	});
}

// ─── Report definition ────────────────────────────────────────────────────────
frappe.query_reports["Tog calculation report"] = {
	filters: [],

	onload: function (report) {
		// Add "Debug Item" button in the report toolbar
		report.page.add_inner_button(__("Debug Item"), function () {
			const dialog = new frappe.ui.Dialog({
				title: __("Debug Item — Tog Calculation"),
				fields: [
					{
						fieldtype: "Link",
						fieldname: "item_code",
						options: "Item",
						label: __("Item Code"),
						reqd: 1,
						description: __("Results will appear in DevTools → Console (F12)"),
					},
				],
				primary_action_label: __("Show in Console"),
				primary_action: function (values) {
					dialog.hide();
					tog_debug_item(values.item_code);
				},
			});
			dialog.show();
		});
	},
};
